File: ecosaras/ecosaras/report/expense_claim_added_filters/expense_claim_added_filters.js
// Copyright (c) 2024, sushant and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Expense claim added filters"] = {
	"filters": [
		{
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_start(),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_end(),
            "reqd": 1
        },
        {
            "fieldname": "from_employee",
            "label": __("From Employee"),
            "fieldtype": "Link",
            "options": "Employee"
        },
        {
            "fieldname": "employee_name",
            "label": __("Employee Name"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },
        {
            "fieldname": "approval_status",
            "label": __("Status"),
            "fieldtype": "Select",
            "default": "Approved",
			"options": [
                { "label": __("All"), "value": "" }, 
				{ "label": __("Draft"), "value": "Draft" }, 
				{ "label": __("Approved"), "value": "Approved" },
				{ "label": __("Rejected"), "value": "Rejected" }
            ]   
        }

	]
};

File: ecosaras/ecosaras/report/expense_claim_added_filters/expense_claim_added_filters.py
import frappe
from frappe import _

def execute(filters=None):
    columns, data = [], []
    
    columns = [
        {"label": _("ID"), "fieldname": "name", "fieldtype": "Link", "options": "Expense Claim", "width": 100},
        {"label": _("From Employee"), "fieldname": "employee", "fieldtype": "Link", "options": "Employee", "width": 150},
        {"label": _("Employee Name"), "fieldname": "employee_name", "fieldtype": "Data", "width": 150},
        {"label": _("Company"), "fieldname": "company", "fieldtype": "Link", "options": "Company", "width": 150},
        {"label": _("Expense Date"), "fieldname": "expense_date", "fieldtype": "Date", "width": 120},
        {"label": _("Expense Type"), "fieldname": "expense_type", "fieldtype": "Data", "width": 150},
        {"label": _("Description"), "fieldname": "description", "fieldtype": "Data", "width": 250},
        {"label": _("Amount"), "fieldname": "amount", "fieldtype": "Currency", "width": 150},  
        {"label": _("Total Claimed Amount"), "fieldname": "total_claimed_amount", "fieldtype": "Currency", "width": 150},
        {"label": _("Total Sanctioned Amount"), "fieldname": "total_sanctioned_amount", "fieldtype": "Currency", "width": 150},
        {"label": _("Posting Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 120},
        {"label": _("Approval Status"), "fieldname": "approval_status", "fieldtype": "Data", "width": 120}
    ]

    conditions = ""
    if filters.get("from_date") and filters.get("to_date"):
        conditions += "AND ec.posting_date BETWEEN %(from_date)s AND %(to_date)s"
    if filters.get("from_employee"):
        conditions += " AND ec.employee = %(from_employee)s"
    if filters.get("employee_name"):
        conditions += " AND ec.employee_name LIKE %(employee_name)s"
    if filters.get("company"):
        conditions += " AND ec.company = %(company)s"
    if filters.get("approval_status"):
        conditions += " AND ec.approval_status = %(approval_status)s"

    data = frappe.db.sql(f"""
        SELECT
            ec.name, ec.employee, ec.employee_name, ec.company, 
            ecd.description, ecd.expense_type, ecd.expense_date, ecd.amount,
            ec.total_claimed_amount, ec.total_sanctioned_amount, ec.posting_date, ec.approval_status
        FROM
            `tabExpense Claim` ec
        LEFT JOIN
            `tabExpense Claim Detail` ecd ON ecd.parent = ec.name
        WHERE
            ec.docstatus < 2
            {conditions}
        ORDER BY
            ec.posting_date DESC
    """, filters, as_dict=1)


    total_claimed_amount = sum(row['total_claimed_amount'] for row in data)
    total_sanctioned_amount = sum(row['total_sanctioned_amount'] for row in data)
    total_amount= sum(row['amount'] for row in data)

    total_row = {
        "name": _("Total"),
        "total_claimed_amount": total_claimed_amount,
        "total_sanctioned_amount": total_sanctioned_amount,
        "amount": total_amount
    }
    data.append(total_row)

    return columns, data
